File: chapters/08-backpropagation/python/tiny_autograd.py
"""Chapter 8 - a tiny automatic differentiation engine, then XOR learned with it.

Four demonstrations:
  1. the chain rule worked example, verified numerically,
  2. backpropagation through the graph L = (a*b + c)^2 - the figure's numbers,
  3. the engine's gradients re-verified with Chapter 3's numerical checker,
  4. a 2-3-1 tanh network trained on XOR: the weights Chapter 7 hand-wired
     are now LEARNED.

Run from the repository root:
    .venv/bin/python chapters/08-backpropagation/python/tiny_autograd.py
"""

import math


class TrackedValue:
    """A number that remembers how it was computed, so gradients can flow back.

    Each TrackedValue stores four things:
      data      - the actual number (the forward-pass value),
      gradient  - dLoss/d(this value), filled in by the backward pass,
      the parent values it was computed from, and
      a small function holding the LOCAL derivative rule of the operation
      that created it (set by __add__, __mul__, tanh below).

    Doing arithmetic on TrackedValues builds the computation graph as a side
    effect; calling .run_backward_pass() on the final value then applies the
    chain rule through the whole graph.
    """

    def __init__(self, data, parent_values=()):
        self.data = data
        self.gradient = 0.0
        self._parent_values = parent_values
        self._propagate_gradient_to_parents = lambda: None

    @staticmethod
    def _wrap_if_plain_number(candidate):
        """Allow mixing plain floats into expressions: they become leaf nodes."""
        return candidate if isinstance(candidate, TrackedValue) else TrackedValue(float(candidate))

    def __add__(self, other_value):
        other_value = TrackedValue._wrap_if_plain_number(other_value)
        result = TrackedValue(self.data + other_value.data, (self, other_value))

        def propagate_addition_gradient():
            # Addition's local derivative is 1 toward both parents, so the
            # incoming gradient passes through unchanged. "+=" (not "=")
            # because a value used in several places collects gradient from
            # every path.
            self.gradient += result.gradient
            other_value.gradient += result.gradient

        result._propagate_gradient_to_parents = propagate_addition_gradient
        return result

    def __mul__(self, other_value):
        other_value = TrackedValue._wrap_if_plain_number(other_value)
        result = TrackedValue(self.data * other_value.data, (self, other_value))

        def propagate_multiplication_gradient():
            # Multiplication's local derivative toward each parent is the
            # OTHER parent's value: d(a*b)/da = b and d(a*b)/db = a.
            self.gradient += other_value.data * result.gradient
            other_value.gradient += self.data * result.gradient

        result._propagate_gradient_to_parents = propagate_multiplication_gradient
        return result

    def __sub__(self, other_value):
        # Built from existing operations (a - b = a + b*(-1)), so subtraction
        # needs no backward rule of its own - the graph handles it.
        return self + TrackedValue._wrap_if_plain_number(other_value) * -1.0

    def tanh(self):
        tanh_of_data = math.tanh(self.data)
        result = TrackedValue(tanh_of_data, (self,))

        def propagate_tanh_gradient():
            # tanh's slope is expressible from its own OUTPUT: 1 - tanh(z)^2.
            self.gradient += (1.0 - tanh_of_data * tanh_of_data) * result.gradient

        result._propagate_gradient_to_parents = propagate_tanh_gradient
        return result

    def run_backward_pass(self):
        """Backpropagation: seed dL/dL = 1, then apply every local rule in
        reverse construction order.

        The ordering is found by a depth-first walk that lists parents before
        children (a topological sort); walking that list backward guarantees a
        node's gradient is complete before it hands gradient to its parents.
        """
        nodes_in_construction_order = []
        already_visited = set()

        def visit_parents_first(value):
            if id(value) not in already_visited:
                already_visited.add(id(value))
                for parent in value._parent_values:
                    visit_parents_first(parent)
                nodes_in_construction_order.append(value)

        visit_parents_first(self)
        self.gradient = 1.0
        for value in reversed(nodes_in_construction_order):
            value._propagate_gradient_to_parents()


def demonstrate_chain_rule():
    """Section 1 of the chapter: dy/dx of y = (2x+1)^2 at x = 1 equals 12."""
    def composed_function(x):
        return (2.0 * x + 1.0) ** 2

    small_step = 1e-3
    numerical_derivative = (composed_function(1.0 + small_step) - composed_function(1.0 - small_step)) / (2 * small_step)
    print("1. Chain rule check: y = (2x+1)^2 at x=1")
    print(f"   by the chain rule: dy/du * du/dx = 6 * 2 = 12")
    print(f"   numerically:       {numerical_derivative:.3f}")


def demonstrate_graph_backpropagation():
    """Sections 2-3: backpropagate L = (a*b + c)^2 and verify every gradient."""
    input_a = TrackedValue(2.0)
    input_b = TrackedValue(3.0)
    input_c = TrackedValue(-1.0)
    product_u = input_a * input_b
    sum_v = product_u + input_c
    loss_l = sum_v * sum_v
    loss_l.run_backward_pass()

    print()
    print("2. Backpropagation through L = (a*b + c)^2 with a=2, b=3, c=-1")
    print(f"   L = {loss_l.data:.0f}")
    print(f"   dL/da = {input_a.gradient:.0f}   dL/db = {input_b.gradient:.0f}   dL/dc = {input_c.gradient:.0f}   (the figure's numbers)")

    # The engine grades itself: every gradient must match the slow-but-sure
    # numerical method from Chapter 3.
    def loss_from_plain_numbers(a, b, c):
        return (a * b + c) ** 2

    small_step = 1e-6
    print("   numerical re-check:", end="")
    for name, gradient, argument_index in (("a", input_a.gradient, 0), ("b", input_b.gradient, 1), ("c", input_c.gradient, 2)):
        arguments_up = [2.0, 3.0, -1.0]
        arguments_down = [2.0, 3.0, -1.0]
        arguments_up[argument_index] += small_step
        arguments_down[argument_index] -= small_step
        numerical_gradient = (loss_from_plain_numbers(*arguments_up) - loss_from_plain_numbers(*arguments_down)) / (2 * small_step)
        print(f"  dL/d{name} = {numerical_gradient:.3f}", end="")
    print()


XOR_INPUTS = [(0.0, 0.0), (0.0, 1.0), (1.0, 0.0), (1.0, 1.0)]
XOR_TARGETS = [-1.0, 1.0, 1.0, -1.0]  # tanh's natural range is (-1, 1)

# Real neural networks start their weights at small RANDOM values, never chosen
# by hand: randomness is what breaks the symmetry between neurons so they can
# specialize instead of all learning the same thing (Chapter 11 covers weight
# initialization in full). So we draw the 13 starting weights randomly here too.
# To keep the run reproducible - and identical to the C port - we seed a tiny,
# fully specified generator rather than the language's built-in random module.
WEIGHT_INITIALIZATION_SEED = 42
INITIAL_WEIGHT_RANGE = 1.0          # each weight starts uniformly random in [-1, +1]


class ReproducibleRandomGenerator:
    """A minimal linear congruential generator (LCG) for reproducible weights.

    Arguments:
        seed: the starting state; the same seed always yields the same sequence.

    Python's own random module would work, but its algorithm is intricate and
    hard to mirror in C. This LCG is a few lines of integer arithmetic, so the C
    port runs the identical sequence and starts from identical weights.
    """

    def __init__(self, seed):
        self._state = seed & 0xFFFFFFFFFFFFFFFF

    def next_uniform(self, low, high):
        """Return the next pseudo-random float, uniform in [low, high)."""
        # Advance the 64-bit state (Numerical Recipes multiplier/increment), then
        # map its top 53 bits to a double in [0, 1) and rescale to [low, high).
        self._state = (self._state * 6364136223846793005 + 1442695040888963407) & 0xFFFFFFFFFFFFFFFF
        unit_interval = (self._state >> 11) * (1.0 / 9007199254740992.0)
        return low + unit_interval * (high - low)


def train_xor_network():
    """Section 5: train the 2-3-1 tanh network on XOR with the engine.

    The training loop is Chapter 5's (forward, loss, gradients, update); the
    only novelty is that the gradient step is now one call to
    run_backward_pass() instead of hand-derived formulas.
    """
    # Draw the 13 starting weights randomly (see ReproducibleRandomGenerator), in
    # a fixed order: 3 hidden neurons, each [weight for x1, weight for x2, bias],
    # then the output neuron [three weights, bias]. The C port draws in the same
    # order from the same generator, so both start identical.
    weight_generator = ReproducibleRandomGenerator(WEIGHT_INITIALIZATION_SEED)

    def random_weight():
        return TrackedValue(weight_generator.next_uniform(-INITIAL_WEIGHT_RANGE, INITIAL_WEIGHT_RANGE))

    hidden_neuron_parameters = [[random_weight() for _ in range(3)] for _ in range(3)]
    output_neuron_parameters = [random_weight() for _ in range(4)]
    all_parameters = [parameter for neuron in hidden_neuron_parameters for parameter in neuron]
    all_parameters += output_neuron_parameters

    def network_forward(first_input, second_input):
        hidden_activations = []
        for weight_for_x1, weight_for_x2, bias in hidden_neuron_parameters:
            weighted_sum = weight_for_x1 * first_input + weight_for_x2 * second_input + bias
            hidden_activations.append(weighted_sum.tanh())
        output_weighted_sum = (
            output_neuron_parameters[0] * hidden_activations[0]
            + output_neuron_parameters[1] * hidden_activations[1]
            + output_neuron_parameters[2] * hidden_activations[2]
            + output_neuron_parameters[3]
        )
        return output_weighted_sum.tanh()

    learning_rate = 0.2
    epochs_to_print = {0, 10, 50, 100, 500, 1000, 2000}
    print()
    print("3. Training the 2-3-1 tanh network on XOR (targets -1 and +1)")
    print("   epoch   loss       predictions for (0,0) (0,1) (1,0) (1,1)")
    for epoch_number in range(2001):
        predictions_this_epoch = []
        total_loss = TrackedValue(0.0)
        for (first_input, second_input), target in zip(XOR_INPUTS, XOR_TARGETS):
            prediction = network_forward(first_input, second_input)
            predictions_this_epoch.append(prediction.data)
            prediction_error = prediction - target
            total_loss = total_loss + prediction_error * prediction_error

        for parameter in all_parameters:
            parameter.gradient = 0.0
        total_loss.run_backward_pass()
        for parameter in all_parameters:
            parameter.data -= learning_rate * parameter.gradient

        if epoch_number in epochs_to_print:
            formatted_predictions = "  ".join(f"{value:+.3f}" for value in predictions_this_epoch)
            print(f"   {epoch_number:>5}   {total_loss.data:.6f}   {formatted_predictions}")

    print("   -> XOR learned: outputs near -1, +1, +1, -1. No hand-wiring this time.")


def main():
    demonstrate_chain_rule()
    demonstrate_graph_backpropagation()
    train_xor_network()


if __name__ == "__main__":
    main()

File: chapters/08-backpropagation/c/tiny_autograd.c
/*
 * Chapter 8 - a tiny automatic differentiation engine in pure C, then XOR
 * learned with it. Full port of the Python example, same numbers.
 *
 * The structural difference from the Python version (explained in the
 * chapter): C has no operator overloading and no closures, so the graph is an
 * ARENA - one flat array of node structs. Each node stores its value, its
 * gradient, which operation created it, and the array indices of its parents.
 * Creation order is already a valid topological order (parents always exist
 * before children), so the backward pass is a plain reverse loop with a
 * switch on the operation type.
 *
 * Build and run from the repository root:
 *     make -C chapters/08-backpropagation/c
 *     ./chapters/08-backpropagation/c/build/tiny_autograd
 */

#include <math.h>
#include <stdint.h>
#include <stdio.h>
#include <stdlib.h>

#define MAXIMUM_GRAPH_NODES 1024

typedef enum {
    OPERATION_LEAF,      /* an input or parameter: nothing to propagate */
    OPERATION_ADD,
    OPERATION_MULTIPLY,
    OPERATION_TANH,
} OperationType;

typedef struct {
    double data;          /* forward-pass value */
    double gradient;      /* dLoss/d(this node), filled by the backward pass */
    OperationType operation;
    int first_parent;     /* index into the arena, -1 if unused */
    int second_parent;
} GraphNode;

static GraphNode graph_nodes[MAXIMUM_GRAPH_NODES];
static int graph_node_count = 0;

/*
 * Create a leaf node (an input or a trainable parameter).
 *
 * data: the number the leaf holds.
 *
 * Returns the new node's arena index. All node-creating functions exit the
 * program if the arena is full - simpler than error codes for a teaching tool.
 */
static int create_leaf(double data) {
    if (graph_node_count >= MAXIMUM_GRAPH_NODES) {
        fprintf(stderr, "graph arena full\n");
        exit(1);
    }
    int node_index = graph_node_count++;
    graph_nodes[node_index] = (GraphNode){data, 0.0, OPERATION_LEAF, -1, -1};
    return node_index;
}

/*
 * Record an addition node: result = nodes[first] + nodes[second].
 *
 * first_parent, second_parent: arena indices of the two values being added.
 */
static int add_values(int first_parent, int second_parent) {
    int node_index = create_leaf(graph_nodes[first_parent].data + graph_nodes[second_parent].data);
    graph_nodes[node_index].operation = OPERATION_ADD;
    graph_nodes[node_index].first_parent = first_parent;
    graph_nodes[node_index].second_parent = second_parent;
    return node_index;
}

/*
 * Record a multiplication node: result = nodes[first] * nodes[second].
 *
 * first_parent, second_parent: arena indices of the two values being multiplied.
 */
static int multiply_values(int first_parent, int second_parent) {
    int node_index = create_leaf(graph_nodes[first_parent].data * graph_nodes[second_parent].data);
    graph_nodes[node_index].operation = OPERATION_MULTIPLY;
    graph_nodes[node_index].first_parent = first_parent;
    graph_nodes[node_index].second_parent = second_parent;
    return node_index;
}

/*
 * Record a tanh node: result = tanh(nodes[parent]).
 *
 * parent: arena index of the value being squashed.
 */
static int tanh_value(int parent) {
    int node_index = create_leaf(tanh(graph_nodes[parent].data));
    graph_nodes[node_index].operation = OPERATION_TANH;
    graph_nodes[node_index].first_parent = parent;
    return node_index;
}

/*
 * Backpropagation: seed dL/dL = 1 at the loss node, then walk the arena in
 * reverse creation order applying each node's local derivative rule.
 *
 * loss_node: arena index of the final value (the loss).
 *
 * Gradients ACCUMULATE with += because a value used in several places
 * collects gradient from every path (the chapter's "one extra rule").
 */
static void run_backward_pass(int loss_node) {
    graph_nodes[loss_node].gradient = 1.0;
    for (int node_index = loss_node; node_index >= 0; node_index--) {
        GraphNode *node = &graph_nodes[node_index];
        switch (node->operation) {
        case OPERATION_LEAF:
            break;
        case OPERATION_ADD:
            /* Addition passes the incoming gradient through unchanged. */
            graph_nodes[node->first_parent].gradient += node->gradient;
            graph_nodes[node->second_parent].gradient += node->gradient;
            break;
        case OPERATION_MULTIPLY:
            /* Each parent's slope is the OTHER parent's value. */
            graph_nodes[node->first_parent].gradient += graph_nodes[node->second_parent].data * node->gradient;
            graph_nodes[node->second_parent].gradient += graph_nodes[node->first_parent].data * node->gradient;
            break;
        case OPERATION_TANH:
            /* tanh's slope from its own output: 1 - tanh(z)^2. */
            graph_nodes[node->first_parent].gradient += (1.0 - node->data * node->data) * node->gradient;
            break;
        }
    }
}

/* ------------------------------------------------------------------ demos */

static double composed_function(double x) {
    return (2.0 * x + 1.0) * (2.0 * x + 1.0);
}

static void demonstrate_chain_rule(void) {
    const double small_step = 1e-3;
    double numerical_derivative =
        (composed_function(1.0 + small_step) - composed_function(1.0 - small_step)) / (2.0 * small_step);
    printf("1. Chain rule check: y = (2x+1)^2 at x=1\n");
    printf("   by the chain rule: dy/du * du/dx = 6 * 2 = 12\n");
    printf("   numerically:       %.3f\n", numerical_derivative);
}

static double loss_from_plain_numbers(double a, double b, double c) {
    return (a * b + c) * (a * b + c);
}

static void demonstrate_graph_backpropagation(void) {
    graph_node_count = 0;
    int input_a = create_leaf(2.0);
    int input_b = create_leaf(3.0);
    int input_c = create_leaf(-1.0);
    int product_u = multiply_values(input_a, input_b);
    int sum_v = add_values(product_u, input_c);
    int loss_l = multiply_values(sum_v, sum_v);
    run_backward_pass(loss_l);

    printf("\n2. Backpropagation through L = (a*b + c)^2 with a=2, b=3, c=-1\n");
    printf("   L = %.0f\n", graph_nodes[loss_l].data);
    printf("   dL/da = %.0f   dL/db = %.0f   dL/dc = %.0f   (the figure's numbers)\n",
           graph_nodes[input_a].gradient, graph_nodes[input_b].gradient, graph_nodes[input_c].gradient);

    /* The engine grades itself against Chapter 3's numerical method. */
    const double small_step = 1e-6;
    printf("   numerical re-check:");
    printf("  dL/da = %.3f",
           (loss_from_plain_numbers(2.0 + small_step, 3.0, -1.0) - loss_from_plain_numbers(2.0 - small_step, 3.0, -1.0)) / (2.0 * small_step));
    printf("  dL/db = %.3f",
           (loss_from_plain_numbers(2.0, 3.0 + small_step, -1.0) - loss_from_plain_numbers(2.0, 3.0 - small_step, -1.0)) / (2.0 * small_step));
    printf("  dL/dc = %.3f\n",
           (loss_from_plain_numbers(2.0, 3.0, -1.0 + small_step) - loss_from_plain_numbers(2.0, 3.0, -1.0 - small_step)) / (2.0 * small_step));
}

/* ------------------------------------------------------------ XOR training */

/* Real networks start their weights at small RANDOM values, never chosen by
 * hand: randomness breaks the symmetry between neurons so they can specialize
 * (Chapter 11 covers weight initialization). So we draw the 13 starting weights
 * randomly here too. This tiny linear congruential generator (LCG) is the exact
 * same one the Python version uses - same seed, same 64-bit arithmetic, same
 * sequence - so both languages start from identical weights. */
#define WEIGHT_INITIALIZATION_SEED 42u
#define INITIAL_WEIGHT_RANGE 1.0

static uint64_t random_generator_state;

static double next_uniform_weight(double low, double high) {
    /* Advance the 64-bit state (unsigned overflow wraps mod 2^64, matching
     * Python's mask), then map its top 53 bits to [0, 1) and rescale. */
    random_generator_state = random_generator_state * 6364136223846793005ULL + 1442695040888963407ULL;
    double unit_interval = (double)(random_generator_state >> 11) * (1.0 / 9007199254740992.0);
    return low + unit_interval * (high - low);
}

static const double xor_inputs[4][2] = {{0, 0}, {0, 1}, {1, 0}, {1, 1}};
static const double xor_targets[4] = {-1.0, 1.0, 1.0, -1.0};

#define PARAMETER_COUNT 13

static void train_xor_network(void) {
    /* The 13 parameters live at arena indices 0..12 for the whole training
     * run; each epoch the arena is reset to just past them, so the forward
     * graph is rebuilt fresh while the parameters persist. */
    graph_node_count = 0;
    int parameter_nodes[PARAMETER_COUNT];
    int parameter_index = 0;
    /* Draw the 13 weights randomly, in the same order as Python: 3 hidden
     * neurons (each [x1 weight, x2 weight, bias]), then the output neuron. */
    random_generator_state = WEIGHT_INITIALIZATION_SEED;
    for (int neuron_index = 0; neuron_index < 3; neuron_index++) {
        for (int component_index = 0; component_index < 3; component_index++) {
            parameter_nodes[parameter_index++] =
                create_leaf(next_uniform_weight(-INITIAL_WEIGHT_RANGE, INITIAL_WEIGHT_RANGE));
        }
    }
    for (int component_index = 0; component_index < 4; component_index++) {
        parameter_nodes[parameter_index++] =
            create_leaf(next_uniform_weight(-INITIAL_WEIGHT_RANGE, INITIAL_WEIGHT_RANGE));
    }

    const double learning_rate = 0.2;
    const int epochs_to_print[] = {0, 10, 50, 100, 500, 1000, 2000};
    printf("\n3. Training the 2-3-1 tanh network on XOR (targets -1 and +1)\n");
    printf("   epoch   loss       predictions for (0,0) (0,1) (1,0) (1,1)\n");

    for (int epoch_number = 0; epoch_number <= 2000; epoch_number++) {
        graph_node_count = PARAMETER_COUNT;
        for (int i = 0; i < PARAMETER_COUNT; i++) {
            graph_nodes[parameter_nodes[i]].gradient = 0.0;
        }

        double predictions_this_epoch[4];
        int total_loss = create_leaf(0.0);
        for (int example_index = 0; example_index < 4; example_index++) {
            int first_input = create_leaf(xor_inputs[example_index][0]);
            int second_input = create_leaf(xor_inputs[example_index][1]);

            /* Forward pass, mirroring the Python network_forward() exactly:
             * three hidden tanh neurons, then one tanh output neuron. */
            int hidden_activations[3];
            for (int neuron_index = 0; neuron_index < 3; neuron_index++) {
                int weight_x1 = parameter_nodes[neuron_index * 3 + 0];
                int weight_x2 = parameter_nodes[neuron_index * 3 + 1];
                int bias = parameter_nodes[neuron_index * 3 + 2];
                int weighted_sum = add_values(
                    add_values(multiply_values(weight_x1, first_input),
                               multiply_values(weight_x2, second_input)),
                    bias);
                hidden_activations[neuron_index] = tanh_value(weighted_sum);
            }
            int output_weighted_sum = add_values(
                add_values(
                    add_values(multiply_values(parameter_nodes[9], hidden_activations[0]),
                               multiply_values(parameter_nodes[10], hidden_activations[1])),
                    multiply_values(parameter_nodes[11], hidden_activations[2])),
                parameter_nodes[12]);
            int prediction = tanh_value(output_weighted_sum);
            predictions_this_epoch[example_index] = graph_nodes[prediction].data;

            /* error = prediction - target, built as prediction + target*(-1)
             * exactly like the Python __sub__ does. */
            int negated_target = create_leaf(-xor_targets[example_index]);
            int prediction_error = add_values(prediction, negated_target);
            total_loss = add_values(total_loss, multiply_values(prediction_error, prediction_error));
        }

        run_backward_pass(total_loss);
        for (int i = 0; i < PARAMETER_COUNT; i++) {
            graph_nodes[parameter_nodes[i]].data -= learning_rate * graph_nodes[parameter_nodes[i]].gradient;
        }

        for (int print_index = 0; print_index < 7; print_index++) {
            if (epochs_to_print[print_index] == epoch_number) {
                printf("   %5d   %.6f   %+.3f  %+.3f  %+.3f  %+.3f\n",
                       epoch_number, graph_nodes[total_loss].data,
                       predictions_this_epoch[0], predictions_this_epoch[1],
                       predictions_this_epoch[2], predictions_this_epoch[3]);
            }
        }
    }
    printf("   -> XOR learned: outputs near -1, +1, +1, -1. No hand-wiring this time.\n");
}

int main(void) {
    demonstrate_chain_rule();
    demonstrate_graph_backpropagation();
    train_xor_network();
    return 0;
}
